scripts/lib/setup-apt-repo

```diff
@@ -46,16 +46,22 @@ deb-src http://apt.postgresql.org/pub/repos/apt/ $release-pgdg main
 deb http://ppa.launchpad.net/groonga/ppa/ubuntu $release main
 deb-src http://ppa.launchpad.net/groonga/ppa/ubuntu $release main
 EOF
+    wget "https://packages.groonga.org/ubuntu/groonga-apt-source-latest-${release}.deb"
+    apt-get -y install "./groonga-apt-source-latest-${release}.deb"
+    rm "./groonga-apt-source-latest-${release}.deb"
+    touch "$STAMP_FILE"
 elif [[ "$release" =~ ^(buster)$ ]]; then
     apt-key add "$SCRIPTS_PATH"/setup/pgdg.asc
-    apt-key add "$SCRIPTS_PATH"/setup/pgroonga-debian.asc
     cat >$SOURCES_FILE <<EOF
 deb http://apt.postgresql.org/pub/repos/apt/ $release-pgdg main
 deb-src http://apt.postgresql.org/pub/repos/apt/ $release-pgdg main
-
-deb https://packages.groonga.org/debian/ $release main
-deb-src https://packages.groonga.org/debian/ $release main
 EOF
+    if ! dpkg-query -l | grep -q groogna-apt-source; then
+        wget "https://packages.groonga.org/debian/groonga-apt-source-latest-${release}.deb"
```

Comment:
Our policy is to verify all our dependencies by cryptographic hash or signature (not just by HTTPS). Also, I don’t think there’s any guarantee that the current directory is writable.

Reply:
Is it OK that PGroonga provides `https://packages.groonga.org/debian/groonga-apt-source-latest-${release}.deb.sha256` and we use it for verifying?

This script is ran as `root`. What situation do you imagine for non-writable case? Where should we use for working directory in this script?

Reply:
> Is it OK that PGroonga provides `https://packages.groonga.org/debian/groonga-apt-source-latest-${release}.deb.sha256` and we use it for verifying?

That path would still be as weak as its weakest link (HTTPS). The solution we previously used on Debian stretch, which had the `groonga-keyring` package, was

1. use our own copy of the key to set up the repository;
2. install the `groonga-keyring` package from the signed repository so the system gets future updates to the key.

I guess we never got around to updating this to use `groonga-archive-keyring`.

> This script is ran as `root`. What situation do you imagine for non-writable case? Where should we use for working directory in this script?

There are reasons `root` might be unable to write to the current directory: it might be on a filesystem that’s read-only or networked or out of space. And there are other reasons it might be a bad idea: it might already contain a file named `groonga-apt-source-latest-${release}.deb` that causes `wget` to unexpectedly write a `.1` file instead, or it might be one that an unprivileged user can overwrite. You should generally only write to a directory if you know what it is.

Reply:
I note that the script already relies on the CWD being readable -- lines 15 and 25, for instance.

Reply:
`touch "$DEPENDENCIES_HASH_FILE"` and `touch "$SOURCES_FILE"`? Those are both absolute paths, not the CWD.

Reply:
The previous solution doesn't work when the Groonga APT repository changes its sign key. We need to update our copy of the sign key. It's not automatic.

`groonga-apt-source` includes the sign key of the Groonga APT repository. And `groonga-apt-source` can be updated by `apt` because the Groonga API repository provides the latest `groonga-apt-source` like `groonga-keyring`.

How about the following?

  * Zulip includes the current sign key of the Groonga APT repository like the current solution
  * PGroonga provides `https://packages.groonga.org/debian/groonga-apt-source-latest-${release}.deb.asc.${FINGERPRINT}` for all sign keys including old sign keys
  * Zulip verify `groonga-apt-source-latest-${release}.deb` with the included sign key
  * Zulip will change the included sign key when PGroonga changes its sign key but it's not required. Because PGroonga also provides `.asc` for old sign keys.

> You should generally only write to a directory if you know what it is.

Can we reuse `/var/lib/zulip`? Or should we create another directory?

Reply:
> The previous solution doesn't work when the Groonga APT repository changes its sign key.

I assume that the new key would be included in the keyring package for a reasonable transitional period before it’s used to re-sign the repository. This is required in any case, to give administrators time to get the new key with `apt upgrade` before they lose the ability to `apt update`.

You’re right, though, that it would break the installer even if it wouldn’t break existing installations.

> How about the following? …

That would work.

> Can we reuse `/var/lib/zulip`? Or should we create another directory?

I suggest creating a new directory with `mktemp -d`, so you know exactly what you’re getting.

Reply:
OK. I'll do them.

Reply:
Done.